`flow.dispatch.workgroup_count_from_slice` surviving due to 0-element tensors.

In some cases there is an `arith.constant` operator with an `index`. This constant fails to lower to the correct vm operation.

Source IR:
https://gist.github.com/rsuderman/bc2b81addc511d0925c94030c73d37ba

Last Pass:
`mlir::iree_compiler::IREE::VM::ConversionPass (iree-vm-conversion)`

IR Immediately before failure:
https://gist.github.com/rsuderman/24d5c02cbfb2ca1fb4c98ebae0057aed

It looks like `index` was never convert to an `i64`. Here is the error.

```
STDERR:
-------
/usr/local/google/home/suderman/Repos/jax/tests/lax_test.py:1918:4: error: 'vm.trunc.i64.i32' op operand #0 must be 64-bit signless integer, but got 'index'
    self._CompileAndCheck(fun, args_maker)
   ^
/usr/local/google/home/suderman/Repos/jax/tests/lax_test.py:1918:4: note: called from
    self._CompileAndCheck(fun, args_maker)
   ^
/usr/local/google/home/suderman/Repos/jax/tests/lax_test.py:1918:4: note: see current operation: %98 = "vm.trunc.i64.i32"(%96#0) : (index) -> i32
    self._CompileAndCheck(fun, args_maker)
   ^
```

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/BUILD.bazel
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/BUILD.bazel
@@ -62,6 +62,7 @@ iree_compiler_cc_library(
         "Passes.cpp",
         "RaiseSpecialOps.cpp",
         "RegionOpUtils.cpp",
+        "RemoveZeroExtentTensors.cpp",
         "SetEncoding.cpp",
         "SplitReduction.cpp",
         "StripAndSplatConstantVariables.cpp",

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/CMakeLists.txt
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/CMakeLists.txt
@@ -61,6 +61,7 @@ iree_cc_library(
     "Passes.cpp"
     "RaiseSpecialOps.cpp"
     "RegionOpUtils.cpp"
+    "RemoveZeroExtentTensors.cpp"
     "SetEncoding.cpp"
     "SplitReduction.cpp"
     "StripAndSplatConstantVariables.cpp"

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp
@@ -215,6 +215,7 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager,
 
   // Preprocessing passes to get the program into a canonical state.
   FunctionLikeNest(passManager)
+      .addPass(IREE::Flow::createRemoveZeroExtentTensorsPass)
       .addPass(IREE::Flow::createDetachElementwiseFromNamedOpsPass)
       .addPass(mlir::createLinalgNamedOpConversionPass)
       .addPass(IREE::Flow::createConvert1X1FilterConv2DToMatmulPass);

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.h
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.h
@@ -127,6 +127,10 @@ std::unique_ptr<Pass> createSetEncodingPass();
 std::unique_ptr<InterfacePass<mlir::FunctionOpInterface>>
 createStripSignednessPass();
 
+// Removes tensors that have 0-extents.
+std::unique_ptr<InterfacePass<mlir::FunctionOpInterface>>
+createRemoveZeroExtentTensorsPass();
+
 // Verifies that the input to the Flow transformation pipeline is legal.
 // This includes checking for operations from dialects that are expected
 // to be legalized before this pass.

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td
@@ -292,6 +292,12 @@ def RaiseSpecialOps :
   let constructor = "mlir::iree_compiler::IREE::Flow::createRaiseSpecialOps()";
 }
 
+def RemoveZeroExtentTensors :
+    InterfacePass<"iree-flow-remove-zero-extent-tensors", "mlir::FunctionOpInterface"> {
+  let summary = "Remove tensors that have 0-extents";
+  let constructor = "mlir::iree_compiler::IREE::Flow::createRemoveZeroExtentTensorsPass()";
+}
+
 def SplitReduction :
     Pass<"iree-flow-split-reduction-ops", ""> {
   let summary = "Split reduction dimension to increase parallelism.";

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/RemoveZeroExtentTensors.cpp
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/RemoveZeroExtentTensors.cpp
@@ -1,0 +1,114 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/MemRef/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+/// Check if a `t` is a `tensor` with zero extents.
+static std::optional<RankedTensorType> isZeroExtent(Type t) {
+  auto operandType = dyn_cast<RankedTensorType>(t);
+  if (operandType &&
+      llvm::any_of(operandType.getShape(), [](int64_t s) { return s == 0; })) {
+    return operandType;
+  }
+  return std::nullopt;
+}
+
+/// Replace operands of the operation that have zero-extent tensors with
+/// a `tensor.empty` op of the same type. This breaks dependencies between
+/// different operations which can be handled subsequently.
+struct ReplaceZeroExtentOperands : public RewritePattern {
+  ReplaceZeroExtentOperands(MLIRContext *context)
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/10, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (isa<tensor::EmptyOp, tensor::DimOp>(op)) {
+      return failure();
+    }
+    Location loc = op->getLoc();
+    bool didUpdate = false;
+    for (OpOperand &operand : op->getOpOperands()) {
+      auto operandType = isZeroExtent(operand.get().getType());
+      if (!operandType) {
+        continue;
+      }
+      if (operand.get().getDefiningOp<tensor::EmptyOp>()) {
+        continue;
+      }
+      Operation *owner = operand.getOwner();
+      int operandNum = operand.getOperandNumber();
+      auto shape = tensor::getMixedSizes(rewriter, loc, operand.get());
+      auto emptyTensorOp = rewriter.create<tensor::EmptyOp>(
+          loc, shape, operandType->getElementType());
+      rewriter.updateRootInPlace(
+          owner, [&]() { owner->setOperand(operandNum, emptyTensorOp); });
+      didUpdate = true;
+    }
+    return success(didUpdate);
+  }
+};
+
+/// Forward the destination of a `tensor.insert_slice` to its uses
+/// if the source is zero-extent.
+struct FoldZeroExtentInserts : public OpRewritePattern<tensor::InsertSliceOp> {
+  using OpRewritePattern<tensor::InsertSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::InsertSliceOp sliceOp,
+                                PatternRewriter &rewriter) const override {
+    if (!isZeroExtent(sliceOp.getSource().getType())) {
+      return failure();
+    }
+    rewriter.replaceOp(sliceOp, sliceOp.getDest());
+    return success();
+  }
+};
+
+namespace {
+
+struct RemoveZeroExtentTensorsPass
+    : RemoveZeroExtentTensorsBase<RemoveZeroExtentTensorsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<tensor::TensorDialect>();
+  }
+  void runOnOperation() override;
+};
+
+} // namespace
+
+void RemoveZeroExtentTensorsPass::runOnOperation() {
+  auto funcOp = getOperation();
+  MLIRContext *context = &getContext();
+  SmallVector<Operation *> opWithZeroExtentTensorOperands;
+  SmallVector<tensor::InsertSliceOp> insertSliceOps;
+
+  RewritePatternSet patterns(context);
+  patterns.insert<FoldZeroExtentInserts, ReplaceZeroExtentOperands>(context);
+  memref::populateResolveRankedShapedTypeResultDimsPatterns(patterns);
+  if (failed(applyPatternsAndFoldGreedily(funcOp, std::move(patterns)))) {
+    funcOp->emitOpError("failed to run canonicalizations (proxy for DCE)");
+    return signalPassFailure();
+  }
+}
+
+std::unique_ptr<InterfacePass<mlir::FunctionOpInterface>>
+createRemoveZeroExtentTensorsPass() {
+  return std::make_unique<RemoveZeroExtentTensorsPass>();
+}
+
+} // namespace Flow
+} // namespace IREE
+} // namespace iree_compiler
+} // namespace mlir

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/BUILD.bazel
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/BUILD.bazel
@@ -48,6 +48,7 @@ iree_lit_test_suite(
             "pad_fusion_with_producer.mlir",
             "pipeline_tests.mlir",
             "raise_special_ops.mlir",
+            "remove_zero_extent_tensors.mlir",
             "set_encoding.mlir",
             "strip_and_splat_constant_variables.mlir",
             "strip_signedness.mlir",

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/CMakeLists.txt
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/CMakeLists.txt
@@ -46,6 +46,7 @@ iree_lit_test_suite(
     "pad_fusion_with_producer.mlir"
     "pipeline_tests.mlir"
     "raise_special_ops.mlir"
+    "remove_zero_extent_tensors.mlir"
     "set_encoding.mlir"
     "strip_and_splat_constant_variables.mlir"
     "strip_signedness.mlir"

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/remove_zero_extent_tensors.mlir
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/remove_zero_extent_tensors.mlir
@@ -1,4 +1,6 @@
-// RUN: iree-opt --iree-flow-remove-zero-extent-tensors --split-input-file %s | FileCheck %s
+// RUN: iree-opt --split-input-file --allow-unregistered-dialect \
+// RUN:          --pass-pipeline="builtin.module(func.func(iree-flow-remove-zero-extent-tensors))" \
+// RUN:          %s | FileCheck %s
 
 func.func @zero_sized_operands(%arg0 : tensor<?x0xf32>, %arg1 : index) -> tensor<?x?xf32> {
   %0 = tensor.empty(%arg1): tensor<0x?xf32>

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/remove_zero_extent_tensors.mlir
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/remove_zero_extent_tensors.mlir
@@ -1,0 +1,36 @@
+// RUN: iree-opt --iree-flow-remove-zero-extent-tensors --split-input-file %s | FileCheck %s
+
+func.func @zero_sized_operands(%arg0 : tensor<?x0xf32>, %arg1 : index) -> tensor<?x?xf32> {
+  %0 = tensor.empty(%arg1): tensor<0x?xf32>
+  %1 = "some_op"(%arg0, %0) : (tensor<?x0xf32>, tensor<0x?xf32>) -> tensor<?x?xf32>
+  return %1 : tensor<?x?xf32>
+}
+//      CHECK: func @zero_sized_operands
+// CHECK-SAME:     %[[ARG0:.+]]: tensor<?x0xf32>
+// CHECK-SAME:     %[[ARG1:.+]]: index
+//      CHECK:   %[[EMPTY0:.+]] = tensor.empty(%[[ARG1]])
+//      CHECK:   %[[DIM:.+]] = tensor.dim %[[ARG0]]
+//      CHECK:   %[[EMPTY1:.+]] = tensor.empty(%[[DIM]])
+//      CHECK:   %[[RESULT:.+]] = "some_op"(%[[EMPTY1]], %[[EMPTY0]]
+//      CHECK:   return %[[RESULT]]
+
+// -----
+
+func.func @zero_sized_tensor_insert(%arg0 : tensor<?x?xf32>, %arg1 : tensor<0x?xf32>,
+    %arg2 : index) -> tensor<?x?xf32> {
+  %1 = tensor.insert_slice %arg1 into %arg0[0, 0] [0, %arg2] [1, 1] : tensor<0x?xf32> into tensor<?x?xf32>
+  return %1 : tensor<?x?xf32>
+}
+// CHECK: func @zero_sized_tensor_insert(%[[ARG0:.+]]: tensor<?x?xf32>
+// CHECK:   return %[[ARG0]] 
+
+// -----
+
+func.func @zero_sizes_tensor_insert_dest(%arg0 : tensor<0x?xf32>, %arg1 : index) -> tensor<0x?xf32> {
+  %0 = tensor.empty(%arg1) : tensor<0x?xf32>
+  %1 = tensor.insert_slice %arg0 into %0[0, 0] [0, %arg1] [1, 1] : tensor<0x?xf32> into tensor<0x?xf32>
+  return %1 : tensor<0x?xf32>
+}
+// CHECK: func @zero_sizes_tensor_insert_dest(%[[ARG0:.+]]: tensor<0x?xf32>, %[[ARG1:.+]]: index)
+// CHECK:   %[[EMPTY:.+]] = tensor.empty(%[[ARG1]])
+// CHECK:   return %[[EMPTY]]

